jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * (C) Copyright 2017-2017, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.tour;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.interfaces.TSPAlgorithm;
+import org.jgrapht.graph.GraphWalk;
+
+import java.util.*;
+
+/**
+ * A dynamic programming algorithm for the TSP problem.
+ *
+ * <p>
+ * The travelling salesman problem (TSP) asks the following question: "Given a list of cities and
+ * the distances between each pair of cities, what is the shortest possible route that visits each
+ * city exactly once and returns to the origin city?".
+ *
+ * <p>
+ * This is an implementation of the Held-Karp algorithm which returns a minimum-cost Hamiltonian tour.
+ * The implementation requires the input graph to contain at least one vertex.
+ * The running time is $O(2^{|V|} \times |V|^2)$ and it takes $O(2^{|V|} \times |V|)$ extra memory.
```

Comment:
Please rephrase the first sentence:
This is an implementation of the Held-Karp algorithm which returns an optimal, minimum-cost Hamiltonian tour.

Just to emphasize that this is an exact algorithm.